### PR TITLE
fix(feat): strip comments during the parsing of crs-setup.conf.example

### PR DESCRIPTION
**Pull request by airween**
This PR modifies the `rules-check.py` script that during the parsing of the `crs-setup.conf.example` it removes the comments before `SecRule` and `SecAction` directives. This means it emulates that rules are valid part of CRS, and - for eg. - therefore it does not need to initialize the TX variables.

The next PR will contain the modification which removes the rules [901167](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/rules/REQUEST-901-INITIALIZATION.conf#L223-L231) and [901169](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/rules/REQUEST-901-INITIALIZATION.conf#L233-L240) from [REQUEST-901-INITIALIZATION.conf](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/rules/REQUEST-901-INITIALIZATION.conf), therefore the `tx.enforce_bodyproc_urlencoded` and `tx.crs_validate_utf8_encoding` variables are no longer initialized.

### CODE REVIEW

File: util/crs-rules-check/rules-check.py [hunk 1]
```diff
@@ -569,6 +569,29 @@ def check_pl_consistency(self):
                     severity  = None     # severity
                     has_nolog = False    # rule has nolog action
 
+def removecomments(data):
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:572]:
```suggestion
def remove_comments(data):
```

File: util/crs-rules-check/rules-check.py [hunk 2]
```diff
@@ -569,6 +569,29 @@ def check_pl_consistency(self):
...
+    _data = []  # new structure by lines
+    lines = [l.strip("\n") for l in data.split("\n")] # split the old content
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:574]: `split` places one entry in the array per delimiter, so there won't be any `\n` left.
```suggestion
    lines = data.split("\n")
```

File: util/crs-rules-check/rules-check.py [hunk 3]
```diff
@@ -569,6 +569,29 @@ def check_pl_consistency(self):
...
+    state = 0   # hold the state of the parser
+    for l in lines:
+        # if the line starts with #SecRule, #SecAction, # SecRule, # SecAction, set the marker
+        if marks.match(l):
+            state = 1
+        # if the marker is set and the line is empty or contains only a comment, unset it
+        if state == 1 and l.strip() in ["", "#"]:
+            state = 0
+
+        # if marker has set, remove the comment
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:585]:
```suggestion
        # if marker is set, remove the comment
```

File: util/crs-rules-check/rules-check.py [hunk 4]
```diff
@@ -630,6 +653,9 @@ def msg(msg):
         try:
             with open(f, 'r') as inputfile:
                 data = inputfile.read()
+                # modify the content of the file, it it is the "crs-setup.conf.example"
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:656]:
```suggestion
                # modify the content of the file, if it is the "crs-setup.conf.example"
```

File: util/crs-rules-check/rules-check.py [hunk 5]
```diff
@@ -694,6 +720,9 @@ def msg(msg):
         try:
             with open(f, 'r') as fp:
                 fromlines = fp.readlines()
+                if f.startswith("crs-setup.conf.example"):
+                    fromlines = removecomments("".join(fromlines)).split("\n")
+                    fromlines = [l + "\n" for l in fromlines]
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:725]: Wouldn't this result in double newlines for each line? `remove_comments` already joins lines with `\n`.

**Review by theseion — CHANGES_REQUESTED** [review]

File: util/crs-rules-check/rules-check.py (hunk 1 above)
Comment (fzipi) [inline util/crs-rules-check/rules-check.py:572]: Can you also add a `"""` comment on what this method is supposed to do? Removing comments where? In general, only inside a `Sec(Rule|Action)`? It's not clear...

File: util/crs-rules-check/rules-check.py [hunk 6]
```diff
@@ -630,6 +653,9 @@ def msg(msg):
         try:
             with open(f, 'r') as inputfile:
                 data = inputfile.read()
+                # modify the content of the file, it it is the "crs-setup.conf.example"
+                if f.startswith("crs-setup.conf.example"):
```
Comment (fzipi) [inline util/crs-rules-check/rules-check.py:657]: Why 'startswith'? Shouldn't be equal to?

File: util/crs-rules-check/rules-check.py [hunk 7]
```diff
@@ -694,6 +720,9 @@ def msg(msg):
         try:
             with open(f, 'r') as fp:
                 fromlines = fp.readlines()
+                if f.startswith("crs-setup.conf.example"):
```
Comment (fzipi) [inline util/crs-rules-check/rules-check.py:723]: Same here.

File: util/crs-rules-check/rules-check.py (hunk 6 above)
Comment (airween) [reply util/crs-rules-check/rules-check.py:657]: In practice I'm using this script that I have many different version of `crs-setup.conf.example`, eg. `crs-setup.conf.example.33`, `crs-setup.conf.example.40`, and so on. With `startswith` I can pass that file without modifying of the target file. It's not a big deal - I can replace it, but if we can live with that, I'd keep this.

File: util/crs-rules-check/rules-check.py (hunk 7 above)
Comment (airween) [reply util/crs-rules-check/rules-check.py:723]: Same as above.

File: util/crs-rules-check/rules-check.py (hunk 1 above)
Comment (airween) [reply util/crs-rules-check/rules-check.py]: Thanks, done - see commit a553627c.

**Comment by airween** [comment]
Thanks guys for your comments, @theseion, @fzipi.

File: util/crs-rules-check/rules-check.py (hunk 5 above)
Comment (theseion) [reply util/crs-rules-check/rules-check.py:762]: Thanks for the explanation.

**Review by theseion — APPROVED** [review]

**Comment by airween** [comment]
@fzipi, there are [three](https://github.com/coreruleset/coreruleset/pull/3161#discussion_r1140151840) [unresolved](https://github.com/coreruleset/coreruleset/pull/3161#discussion_r1140153226) [conversations](https://github.com/coreruleset/coreruleset/pull/3161#discussion_r1140153613) - is there anything what I can explain better?

File: util/crs-rules-check/rules-check.py (hunk 6 above)
Comment (fzipi) [reply util/crs-rules-check/rules-check.py:694]: 🤷

**Comment by dune73** [comment]
Thank you for this welcome workaround. Really appreciated and a good step forward. Talking things through, often again and again, really helps.

**Pull request merged**